src/classes/CW_AudienceCSVUploadController.cls

```diff
@@ -0,0 +1,313 @@
+        Schema.SObjectField agmExternalField = CW_Audience_Group_Member__c.Fields.CompositeKey__c ;
+        
+        for(List<String> flds : csvLines)
+        {
+        	if (flds.get(ConstantUtil.CSV_ACCOUNT_PLAYER_ID_COLUMN)!='')
+        	{
+	            CW_Audience_Group_Member__c aMember = new CW_Audience_Group_Member__c();
+	            // this will need to be looked up from the database
+	            aMember.Audience_Group_LU__c = groupsIdMap.get(flds.get(ConstantUtil.CSV_GROUP_NAME_COLUMN));
+	            // get contact id using the accont name from the csv file
+	            aMember.Contact__c = contactsIdMap.get(flds.get(ConstantUtil.CSV_ACCOUNT_PLAYER_ID_COLUMN));
+	            //create audience group member composite key
+	            aMember.CompositeKey__c = flds.get(ConstantUtil.CSV_ACCOUNT_PLAYER_ID_COLUMN)+flds.get(ConstantUtil.CSV_GROUP_NAME_COLUMN)+scheduleId;
+	            //create ags composite key to get the id of the ags from the agsIdMap
+	            String agsCompositeKey = flds.get(ConstantUtil.CSV_GROUP_NAME_COLUMN)+scheduleId;
+	            aMember.Audience_Group_Schedule__c = agsIdMap.get(agsCompositeKey);
+	            aMembers.add(aMember);
+        	}
+            
+        }
+        csvLines = null;
+		//upsert list of members added to the csv file
+        Database.UpsertResult[] sr = Database.upsert(aMembers,agmExternalField,false);        
+        for(Database.UpsertResult srRecord: sr)
+        {
+            if(srRecord.isSuccess())
+            {
+                System.debug('AGM Id is '+srRecord.getId());
+            }
+            else
+            {
+                for(Database.Error err : srRecord.getErrors()) 
+                {
+                    System.debug('So Error is '+err.getStatusCode() + ': ' + err.getMessage());   
+                    System.debug('Fields that affected this error: ' + err.getFields()); 
+                }
+            }
+        }
+    }
+    
+    /*** CREATE MAPS TO HELP UPLOAD GROUP MEMBERS ***/
+    
+    //create group map to get the group id based on the group name added to the csv file
+	public static Map<String,Id> createGroupsIdMap(Id campaignId)
+    {
+    	groupsIdMap = new Map<String,Id>();
+    	
+        for(CW_Audience_Group__c ag : [select id, Name from CW_Audience_Group__c where Campaign__c = :campaignId])
+        {
+            groupsIdMap.put(ag.Name,ag.Id);
+        }
+        
+        return groupsIdMap;
+    }
+    
+    //create audience group schedule mao based on the group names added to the csv file and schedule id selected before upload
+    public static Map<String,Id> createAgsIdMap(Set<String> groupNames)
+    {
+    	agsIdMap = new Map<String,Id>();
+    	
+        for(CW_Audience_Group_Schedule__c ags : [SELECT ID,CompositeKey__c,Audience_Group__r.Name FROM CW_Audience_Group_Schedule__c where Audience_Group__r.Name In :groupNames])
```

Comment:
Problem: Unbound SOQL query. Reason: a SOQL query can return up to 50,000 records. If this number is exceeded a runtime exception is thrown. Suggestion: cap the max number of records returned adding a LIMIT clause to the query.